src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,160 @@
+            first = new Node(value, null, null);
+            last = first;
+        } else {
+            Node<T> newLast = new Node(value, last, null);
+            last.next = newLast;
+            last = newLast;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == size) {
+            linkLast(value);
+        } else {
+            link(value, node(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return node(index).item;
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        Node<T> node = node(index);
+        node.item = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public T remove(T t) {
+        for (Node<T> node = first; node != null; node = node.next) {
+            if (node.item.equals(t)) {
+                return unlink(node);
+            }
+        }
         return null;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void link(T value, Node<T> x) {
+        Node<T> previous = x.previous;
+        Node<T> newNode = new Node(value, previous, x);
+        x.previous = newNode;
+        if (previous == null) {
+            first = newNode;
+        } else {
+            previous.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkLast(T value) {
+        Node<T> previous = last;
+        Node<T> newNode = new Node(value, previous, null);
+        last = newNode;
+        if (previous == null) {
+            first = newNode;
+        } else {
+            previous.next = newNode;
+        }
+        size++;
+    }
+
+    private T unlink(Node<T> n) {
+        final T value = n.item;
+        final Node<T> next = n.next;
+        final Node<T> previous = n.previous;
+        if (previous == null) {
+            first = next;
+        } else {
+            previous.next = next;
+            n.previous = null;
+        }
+        if (next == null) {
+            last = previous;
+        } else {
+            next.previous = previous;
+            n.next = null;
+        }
+        n.item = null;
+        size--;
+        return value;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private Node<T> node(int index) {
```

Comment:
@joej0e method names should start from verb: getSomethod, addSomethig, doSomething, isSomething, etc...